File: tl/src/main/java/com/github/badoualy/telegram/tl/api/TLUpdateContactLink.java
package com.github.badoualy.telegram.tl.api;

import com.github.badoualy.telegram.tl.TLContext;

import java.io.IOException;
import java.io.InputStream;
import java.io.OutputStream;

import static com.github.badoualy.telegram.tl.StreamUtils.readInt;
import static com.github.badoualy.telegram.tl.StreamUtils.readTLObject;
import static com.github.badoualy.telegram.tl.StreamUtils.writeInt;
import static com.github.badoualy.telegram.tl.StreamUtils.writeTLObject;
import static com.github.badoualy.telegram.tl.TLObjectUtils.SIZE_CONSTRUCTOR_ID;
import static com.github.badoualy.telegram.tl.TLObjectUtils.SIZE_INT32;

/**
 * @author Yannick Badoual yann.badoual@gmail.com
 * @see <a href="http://github.com/badoualy/kotlogram">http://github.com/badoualy/kotlogram</a>
 */
public class TLUpdateContactLink extends TLAbsUpdate {

    public static final int CONSTRUCTOR_ID = 0x9d2e67c5;

    protected int userId;

    protected TLAbsContactLink myLink;

    protected TLAbsContactLink foreignLink;

    private final String _constructor = "updateContactLink#9d2e67c5";

    public TLUpdateContactLink() {
    }

    public TLUpdateContactLink(int userId, TLAbsContactLink myLink, TLAbsContactLink foreignLink) {
        this.userId = userId;
        this.myLink = myLink;
        this.foreignLink = foreignLink;
    }

    @Override
    public void serializeBody(OutputStream stream) throws IOException {
        writeInt(userId, stream);
        writeTLObject(myLink, stream);
        writeTLObject(foreignLink, stream);
    }

    @Override
    @SuppressWarnings({"unchecked", "SimplifiableConditionalExpression"})
    public void deserializeBody(InputStream stream, TLContext context) throws IOException {
        userId = readInt(stream);
        myLink = readTLObject(stream, context, TLAbsContactLink.class, -1);
        foreignLink = readTLObject(stream, context, TLAbsContactLink.class, -1);
    }

    @Override
    public int computeSerializedSize() {
        int size = SIZE_CONSTRUCTOR_ID;
        size += SIZE_INT32;
        size += myLink.computeSerializedSize();
        size += foreignLink.computeSerializedSize();
        return size;
    }

    @Override
    public String toString() {
        return _constructor;
    }

    @Override
    public int getConstructorId() {
        return CONSTRUCTOR_ID;
    }

    public int getUserId() {
        return userId;
    }

    public void setUserId(int userId) {
        this.userId = userId;
    }

    public TLAbsContactLink getMyLink() {
        return myLink;
    }

    public void setMyLink(TLAbsContactLink myLink) {
        this.myLink = myLink;
    }

    public TLAbsContactLink getForeignLink() {
        return foreignLink;
    }

    public void setForeignLink(TLAbsContactLink foreignLink) {
        this.foreignLink = foreignLink;
    }
}

File: tl/src/main/kotlin/com/github/badoualy/telegram/tl/api/TLUpdateContactLink.kt
package com.github.badoualy.telegram.tl.api

import com.github.badoualy.telegram.tl.TLObjectUtils.SIZE_BOOLEAN
import com.github.badoualy.telegram.tl.TLObjectUtils.SIZE_CONSTRUCTOR_ID
import com.github.badoualy.telegram.tl.TLObjectUtils.SIZE_DOUBLE
import com.github.badoualy.telegram.tl.TLObjectUtils.SIZE_INT32
import com.github.badoualy.telegram.tl.TLObjectUtils.SIZE_INT64
import com.github.badoualy.telegram.tl.TLObjectUtils.computeTLBytesSerializedSize
import com.github.badoualy.telegram.tl.TLObjectUtils.computeTLStringSerializedSize
import com.github.badoualy.telegram.tl.serialization.TLDeserializer
import com.github.badoualy.telegram.tl.serialization.TLSerializer
import java.io.IOException
import kotlin.Any
import kotlin.Boolean
import kotlin.Int
import kotlin.String
import kotlin.jvm.Throws

/**
 * updateContactLink#9d2e67c5
 *
 * @author Yannick Badoual yann.badoual@gmail.com
 * @see <a href="http://github.com/badoualy/kotlogram">http://github.com/badoualy/kotlogram</a>
 */
class TLUpdateContactLink() : TLAbsUpdate() {
    var userId: Int = 0

    var myLink: TLAbsContactLink = TLContactLinkHasPhone()

    var foreignLink: TLAbsContactLink = TLContactLinkHasPhone()

    private val _constructor: String = "updateContactLink#9d2e67c5"

    override val constructorId: Int = CONSTRUCTOR_ID

    constructor(
            userId: Int,
            myLink: TLAbsContactLink,
            foreignLink: TLAbsContactLink
    ) : this() {
        this.userId = userId
        this.myLink = myLink
        this.foreignLink = foreignLink
    }

    @Throws(IOException::class)
    override fun serializeBody(tlSerializer: TLSerializer) = with (tlSerializer)  {
        writeInt(userId)
        writeTLObject(myLink)
        writeTLObject(foreignLink)
    }

    @Throws(IOException::class)
    override fun deserializeBody(tlDeserializer: TLDeserializer) = with (tlDeserializer)  {
        userId = readInt()
        myLink = readTLObject<TLAbsContactLink>()
        foreignLink = readTLObject<TLAbsContactLink>()
    }

    override fun computeSerializedSize(): Int {
        var size = SIZE_CONSTRUCTOR_ID
        size += SIZE_INT32
        size += myLink.computeSerializedSize()
        size += foreignLink.computeSerializedSize()
        return size
    }

    override fun toString() = _constructor

    override fun equals(other: Any?): Boolean {
        if (other !is TLUpdateContactLink) return false
        if (other === this) return true

        return userId == other.userId
                && myLink == other.myLink
                && foreignLink == other.foreignLink
    }
    companion object  {
        const val CONSTRUCTOR_ID: Int = 0x9d2e67c5.toInt()
    }
}
